### ISSUE TITLE
prompt buffer: ignore 'modified' state

### ISSUE THREAD

**Issue by mfussenegger** [neovim/neovim#14116]
[nvim-dap](https://github.com/mfussenegger/nvim-dap) uses `buftype=prompt` for the debug REPL.

It looks as if by default any input to a prompt buffer sets the `modified` flag. This is a bit annoying for the REPL case, because users usually don't care whether the REPL is modified or not and they shouldn't get a "Did not save [dap-repl]" message. So I added something like this:

```
api.nvim_buf_attach(buf, false, {
  on_lines = function(b)
    api.nvim_buf_set_option(b, 'modified', false)
  end;
  on_changedtick = function(b)
    api.nvim_buf_set_option(b, 'modified', false)
  end;
  ...
```

But this workaround had the side effect of [breaking the backspace key](https://github.com/mfussenegger/nvim-dap/issues/119) in that after you type something you can no longer remove the characters with backspace.

Is there any other good way to that would allow to treat a prompt buffer like a terminal or `nofile` where modifications don't matter?

**Comment by bfredl** [comment]
probably `buftype=prompt,nofile` should be allowed. Perhaps it can be suggested upstream?

**Comment by cogig** [comment]
Are there any updates?

**Comment by mfussenegger** [comment]
Afaik no one has reported it upstream to vim yet. I myself didn't get to it - because it would first be necessary to verify that the behavior isn't already different and the problem also occurs within vim.

If you could take care of that, that would be great.

**Comment by cogig** [comment]
@mfussenegger I have tried the latest version of neovim.
```
NVIM v0.6.0-dev+65-g46009499a
Build type: Release
LuaJIT 2.1.0-beta3
```
The problem still exists in nvim-dap repl.

**Comment by cogig** [comment]
@mfussenegger @bfredl How to comfirm that the problem also occurs within vim? I have no idea. Can you give me any advise?

**Comment by pittcat** [comment]
@mfussenegger @bfredl I think the bug has large degree of influence on usage of repl and watch window. Many key can't work well.

**Comment by pfeiferj** [comment]
I'm currently using the following vimscript as a workaround, it's not fool proof but it probably works for most people. The biggest issue that I know of is that you won't be able to backspace anywhere there is a "> " but if someone felt like updating it you could just compare to the position of the first "> " instead of the current method that's just looking at the character under the cursor and the character before the cursor to fix that. It's also probably not the cleanest vimscript ever and I would have used lua instead but it turns out that if you try to run lua on the OptionSet autocommand it causes errors when opening the :help text and possibly other areas too.

```vim
fun! PromptBackspace()
  let currentPos = getline('.')[col('.')-1]
  let beforePos = getline('.')[col('.')-2]
  if !(currentPos == " " && beforePos == ">")
    if col(".") == col("$")-1
      normal x
      startinsert!
    else
      normal x
      startinsert
    endif
  else
    startinsert!
  endif
endfun

fun! PromptBackspaceSetup()
    if v:option_new == 'prompt'
      inoremap <buffer> <backspace> <esc>:call PromptBackspace()<cr>
    endif
endfun

autocmd OptionSet * call PromptBackspaceSetup()
```

**Comment by gpanders** [comment]
I ran into this myself a little while ago while working on a Fennel REPL plugin. Vim modified the behavior of `'modified'` to respect `buftype=prompt` in https://github.com/vim/vim/commit/4551c0a9fcdbdef52836d4852686d54b5e47fdaf, solely for the termdebug plugin.

We could consider breaking with upstream on this and making `buftype=prompt` always ignore `'modified'` as it did before that commit (which I would argue is the less surprising behavior).

**Comment by pfeiferj** [comment]
It seems it's pretty common for plugins to use the workaround mentioned in the original post leading to the broken backspace issue also mentioned in the original post. I ran into it using a couple different plugins and am currently using the following workaround. I decided to just go ahead and set the remap for all prompts since it's a common issue but the logic should be able to be pulled down into individual plugins. That being said, I do think this is a case where we should break from vim in the name of usability and change the default behavior in neovim so these workarounds aren't needed.

```lua
function PromptBackspace() --Expects to have entered normal mode before entering function, will return to insert mode
  local currentLineNumber = vim.fn["line"](".")
  local currentColumnNumber = vim.fn["col"](".")
  local currentLineLength = vim.fn["col"]("$")
  local currentLine = vim.api.nvim_buf_get_lines(0, currentLineNumber-1, currentLineNumber, false)[1]
  local _, endPrompt = string.find(currentLine, "> ")

  if endPrompt == nil then
    endPrompt = 0
  end

  if currentColumnNumber ~= endPrompt then --Not beginning of the prompt
    if currentColumnNumber == currentLineLength - 1 then --Not beginning of prompt and is the end of the line
      vim.cmd([[normal x]])
      vim.cmd([[startinsert!]])
    else --Not beginning of prompt and is not the end of the line
      vim.cmd([[normal x]])
      vim.cmd([[startinsert]])
    end
  else --Beginning of prompt
    vim.cmd([[startinsert]])
  end
end

vim.cmd([[
fun! PromptBackspaceSetup()
    if v:option_new == 'prompt'
      inoremap <buffer> <backspace> <esc>:lua PromptBackspace()<cr>
    endif
endfun

autocmd OptionSet * call PromptBackspaceSetup() "not using a lua function here because of error when accessing :help
]])
```

**Comment by pfeiferj** [comment]
An improvement to the previous backspace workaround that doesn't switch between normal and insert modes:

```lua
function PromptBackspace()
  -- Allows backspacing through previously set text when in a prompt.
  --
  -- Note 1: nvim_buf_[get|set]_lines is 0 indexed vs vim [line|col](".") being 1 indexed
  -- Note 2: Insert mode cursor is after (+1) the column as opposed to in normal mode it would be on (+0) the column.
  -- Note 3: Awkwardly nvim_win_set_cursor is 1 indexed for line and 0 indexed for column.

  local currentLineNumber = vim.fn["line"](".") 
  local currentColumnNumber = vim.fn["col"](".")
  local currentLine = vim.api.nvim_buf_get_lines(0, currentLineNumber-1, currentLineNumber, false)[1]
  local _, endPromptPrefix = string.find(currentLine, "> ")

  if (currentColumnNumber-1) ~= endPromptPrefix then
    vim.api.nvim_buf_set_text(0, currentLineNumber-1, currentColumnNumber-2, currentLineNumber-1, currentColumnNumber-1, {""})
    vim.api.nvim_win_set_cursor(0, {currentLineNumber, currentColumnNumber-2})
  end
end

vim.cmd([[
fun! PromptBackspaceSetup()
    if v:option_new == 'prompt'
      inoremap <buffer> <backspace> <Cmd>lua PromptBackspace()<cr>
    endif
endfun

autocmd OptionSet * call PromptBackspaceSetup() "not using a lua function here because of error when accessing :help
]])
```

**Comment by rcarriga** [comment]
Just FYI @pfeiferj (or anyone using the workaround) `string.len` doesn't work with unicode characters in the prompt, you can use `vim.str_utfindex` instead :smile:

**Comment by pfeiferj** [comment]
whoops, updated the post. Dang unicode messing everything up, clearly the world should have adapted to 256 characters not programmers to languages/emojis lol.

**Comment by beardedsakimonkey** [comment]
> We could consider breaking with upstream on this and making `buftype=prompt` always ignore `'modified'` as it did before that commit (which I would argue is the less surprising behavior).

Does this sound good to other folks?  If so, I'd be happy to take a shot at implementing it if contributions are welcome.

**Comment by brianhuster** [comment]
> Afaik no one has reported it upstream to vim yet

I have reported it upstream and it has been closed as "Not planned" https://github.com/vim/vim/issues/18942

**Comment by justinmk** [comment]
> Vim modified the behavior of `'modified'` to respect `buftype=prompt` in [vim/vim@4551c0a](https://github.com/vim/vim/commit/4551c0a9fcdbdef52836d4852686d54b5e47fdaf), solely for the termdebug plugin.
> 
> We could consider breaking with upstream on this and making `buftype=prompt` always ignore `'modified'` as it did before that commit

Yes let's do this. In fact this was part of my todo list in https://github.com/neovim/neovim/issues/34561 : 

> editing a prompt marks the buffer as "modified", this is usually just unwanted and annoying

**Comment by justinmk** [comment]
> nvim-dap since changed the approach to set `modified=false` only in `ExitPre`

https://github.com/vim/vim/issues/18942#issuecomment-3777877019 mentions `QuitPre` instead. Does that resolve all the quirks @mfussenegger ? If so, perhaps we should just document that instead of reverting https://github.com/neovim/neovim/commit/aec3d7915c55fc0b7dc73f6186cf0ae45d416d33   (https://github.com/neovim/neovim/pull/38118).

**Comment by mfussenegger** [comment]
> > nvim-dap since changed the approach to set `modified=false` only in `ExitPre`
> 
> [vim/vim#18942 (comment)](https://github.com/vim/vim/issues/18942#issuecomment-3777877019) mentions `QuitPre` instead. Does that resolve all the quirks [@mfussenegger](https://github.com/mfussenegger) ? If so, perhaps we should just document that instead of reverting [aec3d79](https://github.com/neovim/neovim/commit/aec3d7915c55fc0b7dc73f6186cf0ae45d416d33) ([#38118](https://github.com/neovim/neovim/pull/38118)).

I haven't tried `QuitPre` as I haven't noticed any issues with `ExitPre`.
